kernel/memory_structs/src/lib.rs

```diff
@@ -266,87 +315,145 @@ macro_rules! implement_page_frame {
 
             #[doc = "A `" $TypeName "` is a chunk of **" $desc "** memory aligned to a [`PAGE_SIZE`] boundary."]
             #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
-            pub struct $TypeName {
+            pub struct $TypeName<P: PageSize = Page4KiB> {
                 number: usize,
+                pub size: PhantomData::<P>,
             }
-
             impl $TypeName {
-                #[doc = "Returns the `" $address "` at the start of this `" $TypeName "`."]
-                pub const fn start_address(&self) -> $address {
-                    $address::new_canonical(self.number * PAGE_SIZE)
+                #[doc = "Returns the `" $TypeName "` containing the given `" $address "`."]
+                pub const fn containing_address(addr: $address) -> $TypeName {
+                    $TypeName {
+                        number: addr.value() / PAGE_SIZE,
+                        size: PhantomData::<Page4KiB>,
+                    }
```

Comment:
you can still make this generic, and combine it with both `containing_address_2mb` and `containing_address_1gb` into one single function.

Just put in the `impl<P: PageSize> $TypeName<P>` block down below, and then use the `P::SIZE` and other constants for the math to calculate the `number` correctly.

Also, with that, you shouldn't need to specify anything in the PhantomData type; I _think_ you can just do:
```rust
    size: PhantomData,
```
in the constructor
```rust

Reply:
In this way, there wouldn't be separate/dedicated functions for the 2mb and 1gb versions. The caller of this function would just specify what kind of page they want with type annotations, which is canonically Rusty:
```rust
let huge_2mb_page = Page::<Page2MiB>::containing_address(...);
let huge_1gb_page = Page::<Page1GiB>::containing_address(...);
```